### PR TITLE
Update/nextjs 12

### PR DESCRIPTION
**Pull request by luiztiago**
## Next.js 12 Summary

- Rust Compiler: ~3x faster Fast Refresh and ~5x faster builds
- Middleware (beta): Enabling full flexibility in Next.js with code over configuration
- React 18 Support: Native Next.js APIs are now supported, as well as Suspense
- <Image /> AVIF Support: Opt-in for 20% smaller images
- Bot-aware ISR Fallback: Optimized SEO for web crawlers
- Native ES Modules Support: Aligning with the standardized module system
- URL Imports (alpha): Import packages from any URL, no installs required
- React Server Components (alpha): Try it today, including SSR streaming

## Other Improvements

- Adding `pages/_app.js` or `pages/_document.js` to the application now automatically replaces the built-in version without requiring a reboot of the Next.js CLI.
- Officially removed webpack 4, but webpack 5 was already the default on Next.js 11.
- The minimum Node.js version has been bumped from `12.0.0` to `12.22.0`, which is the first version of Node.js with native ES modules support.
- The upgrade to 12.0.5+ is recommended for all users. A backport of the patch to Next.js 11 is available as 11.1.3.
> When a URL is provided to `next-server` that cannot be parsed, an `unhandledPromiseRejection` could occur. On Node.js versions < `v15.0.0` this isn't a fatal issue as only a warning is shown. However, in Node.js versions > `v15.0.0`, this causes the server process to exit, which can result in unexpected server crashes.

#### srcSet issue

VIP Preflight was returning an error because the `<Image />` component from Next.js [added the `srcSet` as an exception](https://nextjs.org/docs/api-reference/next/image#other-props) to the props. I removed it for now, and added a comment explaining how to use `srcSet` on `` HTML tag.

```console
➜  vip-go-nextjs-skeleton git:(trunk) ✗ npm run vip:preflight

> vip-go-nextjs-boilerplate@0.1.0 vip:preflight
> npx @automattic/vip-go-preflight-checks


  Welcome to
   _    __________     ______
  | |  / /  _/ __ \   / ____/___
  | | / // // /_/ /  / / __/ __ \
  | |/ // // ____/  / /_/ / /_/ /
  |___/___/_/       \____/\____/
  Preflight Checks for Node Apps

  Running checks for the vip-go-nextjs-boilerplate app...


  [ Step 1/2 ] Checking npm scripts in the project...
  [ Step 1/2 ] Step details: Checking your package.json for npm build and npm start
  [ Step 1/2 ] Step successful 👍


  [ Step 2/2 ] Installing dependencies, building, and starting your app
  [ Step 2/2 ] Step details: Checking if your app accepts a PORT and responds to /cache-healthcheck?
  Info: Installing dependencies using npm install --production...
  Info: Building the project using npm run build...
  Error: Error: Command failed with exit code 1: npm run build
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

Failed to compile.

./components/PostContent/PostContent.tsx:25:10
Type error: Type '{ src: string; srcSet: string; width: string; height: string; alt: string; key: string; }' is not assignable to type 'IntrinsicAttributes & Omit<DetailedHTMLProps<ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement>, "loading" | ... 5 more ... | "height"> & { ...; }'.
  Property 'srcSet' does not exist on type 'IntrinsicAttributes & Omit<DetailedHTMLProps<ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement>, "loading" | ... 5 more ... | "height"> & { ...; }'.

  23 | 								<Image
  24 | 									src={blockProps.src}
> 25 | 									srcSet={blockProps.srcset}
     | 									^
  26 | 									width={blockProps.width || blockProps.originalWidth}
  27 | 									height={blockProps.height || blockProps.originalHeight}
  28 | 									alt={blockProps.alt}

> vip-go-nextjs-boilerplate@0.1.0 prebuild
> npm run codegen


> vip-go-nextjs-boilerplate@0.1.0 codegen
> graphql-codegen --config codegen.yml -r dotenv/config

[22:17:54] Parse configuration [started]
[22:17:54] Parse configuration [completed]
[22:17:54] Generate outputs [started]
[22:17:54] Generate graphql/generated/index.tsx [started]
[22:17:54] Generate graphql/generated/fragmentMatcher.ts [started]
[22:17:54] Load GraphQL schemas [started]
[22:17:54] Load GraphQL schemas [started]
[22:17:56] Load GraphQL schemas [completed]
[22:17:56] Load GraphQL documents [started]
[22:17:56] Load GraphQL documents [completed]
[22:17:56] Generate [started]
[22:17:56] Generate [completed]
[22:17:56] Generate graphql/generated/index.tsx [completed]
[22:17:56] Load GraphQL schemas [completed]
[22:17:56] Load GraphQL documents [started]
[22:17:56] Load GraphQL documents [completed]
[22:17:56] Generate [started]
[22:17:56] Generate [completed]
[22:17:56] Generate graphql/generated/fragmentMatcher.ts [completed]
[22:17:56] Generate outputs [completed]

> vip-go-nextjs-boilerplate@0.1.0 build
> next build

info  - Loaded env from /Users/luiztiago/dev/vip-go-nextjs-skeleton/.env
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
info  - Checking validity of types...
  [ Step 2/2 ] Step failed 😱

Oups! Looks like you need to fix some steps to make your app ready for VIP Go.
```

Running the tests, it's still being passed to the `` tag, but we're getting an error on build to this. To prevent the error, I've created a const named `imageProps` to have the `srcSet` as a key, and then build and validation are working well.

I've tried removing the `srcSet` prop, but on tests, the Network opened the wrong images, setting only the `w=` query param based on the `deviceSizes` setting on the nextjs configuration file. So to have the images loading different URLs as expected from the API, I kept the `srcSet` prop, and it's working as expected.

#### Alt attribute

```console
➜  vip-go-nextjs-skeleton git:(update/nextjs-12) ✗ npm run lint

> vip-go-nextjs-boilerplate@0.1.0 lint
> next lint --ignore-path .gitignore

info  - Loaded env from /Users/luiztiago/dev/vip-go-nextjs-skeleton/.env
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

Why you should do it regularly:
https://github.com/browserslist/browserslist#browsers-data-updating

./components/PostContent/PostContent.tsx
32:9  Warning: Image elements must have an alt prop, either with meaningful text, or an empty string for decorative images.  jsx-a11y/alt-text

info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

```

To prevent this warning, I've added the `alt` as a prop on `<Image />` tag instead of inside `imageProps`.

#### Changed wrapping element

Another update on `next/image` is about the [wrapping element](https://nextjs.org/docs/upgrading#nextimage-changed-wrapping-element). The component now renders the `` inside a `<span>` instead of `<div>`. But the generated `<span>` has a `display: block`, and all the tests work as expected.

##### next/image v11.0

```html
<div style="display:block;overflow:hidden;position:relative;box-sizing:border-box;margin:0"><div style="display:block;box-sizing:border-box;padding-top:46.52014652014652%"></div><img alt="" src="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75" decoding="async" data-nimg="true" style="position:absolute;top:0;left:0;bottom:0;right:0;box-sizing:border-box;padding:0;border:none;margin:auto;display:block;width:0;height:0;min-width:100%;max-width:100%;min-height:100%;max-height:100%" sizes="100vw" srcset="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=640&amp;q=75 640w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=750&amp;q=75 750w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=828&amp;q=75 828w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1080&amp;q=75 1080w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1200&amp;q=75 1200w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1920&amp;q=75 1920w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=2048&amp;q=75 2048w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75 3840w"></div>
```

##### next/image v12.0.4

```html
<span style="box-sizing:border-box;display:block;overflow:hidden;width:initial;height:initial;background:none;opacity:1;border:0;margin:0;padding:0;position:relative"><span style="box-sizing:border-box;display:block;width:initial;height:initial;background:none;opacity:1;border:0;margin:0;padding:0;padding-top:46.52014652014652%"></span><img alt="" src="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75" decoding="async" data-nimg="responsive" style="position:absolute;top:0;left:0;bottom:0;right:0;box-sizing:border-box;padding:0;border:none;margin:auto;display:block;width:0;height:0;min-width:100%;max-width:100%;min-height:100%;max-height:100%" sizes="100vw" srcset="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=640&amp;q=75 640w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=750&amp;q=75 750w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=828&amp;q=75 828w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1080&amp;q=75 1080w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1200&amp;q=75 1200w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1920&amp;q=75 1920w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=2048&amp;q=75 2048w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75 3840w"><noscript><img alt="" sizes="100vw" srcSet="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=640&amp;q=75 640w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=750&amp;q=75 750w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=828&amp;q=75 828w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1080&amp;q=75 1080w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1200&amp;q=75 1200w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=1920&amp;q=75 1920w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=2048&amp;q=75 2048w, https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75 3840w" src="https://wpvip-com-decoupled.go-vip.net/docs/wp-content/uploads/sites/2/2021/05/image-14.png?w=3840&amp;q=75" decoding="async" data-nimg="responsive" style="position:absolute;top:0;left:0;bottom:0;right:0;box-sizing:border-box;padding:0;border:none;margin:auto;display:block;width:0;height:0;min-width:100%;max-width:100%;min-height:100%;max-height:100%" loading="lazy"/></noscript></span>
```

# Testing PR

## Lint

```console
➜  vip-go-nextjs-skeleton git:(update/nextjs-12) ✗ npm run lint

> vip-go-nextjs-boilerplate@0.1.0 lint
> next lint --ignore-path .gitignore

info  - Loaded env from /Users/luiztiago/dev/vip-go-nextjs-skeleton/.env
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

Why you should do it regularly:
https://github.com/browserslist/browserslist#browsers-data-updating
✔ No ESLint warnings or errors

```

## Preflight

```console
➜  vip-go-nextjs-skeleton git:(update/nextjs-12) ✗ npm run vip:preflight                  

> vip-go-nextjs-boilerplate@0.1.0 vip:preflight
> npx @automattic/vip-go-preflight-checks


  Welcome to
   _    __________     ______
  | |  / /  _/ __ \   / ____/___
  | | / // // /_/ /  / / __/ __ \
  | |/ // // ____/  / /_/ / /_/ /
  |___/___/_/       \____/\____/
  Preflight Checks for Node Apps

  Running checks for the vip-go-nextjs-boilerplate app...


  [ Step 1/2 ] Checking npm scripts in the project...
  [ Step 1/2 ] Step details: Checking your package.json for npm build and npm start
  [ Step 1/2 ] Step successful 👍


  [ Step 2/2 ] Installing dependencies, building, and starting your app
  [ Step 2/2 ] Step details: Checking if your app accepts a PORT and responds to /cache-healthcheck?
  Info: Installing dependencies using npm install --production...
  Info: Building the project using npm run build...
  Info: Running your app with npm start on PORT: 3254...
  Info: Sending a GET request to http://localhost:3254/cache-healthcheck?...
  [ Step 2/2 ] Step successful 👍


Congratulations! Your application is ready for VIP Go!
```

## Audit

```console
➜  vip-go-nextjs-skeleton git:(update/nextjs-12) ✗ npm audit fix
npm WARN ERESOLVE overriding peer dependency

added 40 packages, removed 112 packages, changed 6 packages, and audited 1311 packages in 16s

110 packages are looking for funding
  run `npm fund` for details

# npm audit report

ansi-regex  >2.1.1 <5.0.1
Severity: moderate
 Inefficient Regular Expression Complexity in chalk/ansi-regex - https://github.com/advisories/GHSA-93q8-gq69-wqmw
fix available via `npm audit fix --force`
Will install @graphql-codegen/cli@1.9.0, which is a breaking change
node_modules/log-update/node_modules/ansi-regex
  strip-ansi  4.0.0 - 5.2.0
  Depends on vulnerable versions of ansi-regex
  node_modules/log-update/node_modules/strip-ansi
    string-width  2.1.0 - 4.1.0
    Depends on vulnerable versions of strip-ansi
    node_modules/log-update/node_modules/string-width
      wrap-ansi  3.0.0 - 6.1.0
      Depends on vulnerable versions of string-width
      Depends on vulnerable versions of strip-ansi
      node_modules/log-update/node_modules/wrap-ansi
        log-update  2.1.0 - 3.4.0
        Depends on vulnerable versions of wrap-ansi
        node_modules/log-update
          listr-update-renderer  >=0.5.0
          Depends on vulnerable versions of log-update
          node_modules/listr-update-renderer
            @graphql-codegen/cli  <=1.8.4-alpha-ff40af04.41 || >=1.9.1-alpha-01fa8c2d.10
            Depends on vulnerable versions of listr-update-renderer
            node_modules/@graphql-codegen/cli
            listr  >=0.14.3
            Depends on vulnerable versions of listr-update-renderer
            node_modules/listr

8 moderate severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force
```

### Research

- https://nextjs.org/blog/next-12
- https://nextjs.org/docs/upgrading#swc-replacing-terser-for-minification
- https://developer.wordpress.org/apis/handbook/responsive-images/
- https://github.com/vercel/next.js/issues/18413
- https://alistapart.com/article/responsive-images-in-practice/
- https://github.com/vercel/next.js/releases/tag/v12.0.5

### CODE REVIEW

**Comment by chriszarate** [comment]
@luiztiago Could you also run `npm audit fix` as part of this upgrade?

**Comment by luiztiago** [comment]
@chriszarate - I just added the `npm audit fix` to the PR description.

If I run the `npm audit fix --force`, it's downgrading the `@graphql-codegen/cli`. FYI, It's also happening on `trunk` branch. If you want, I can add it to my PR.
 
```console
-    "@graphql-codegen/cli": "^2.1.0",
+    "@graphql-codegen/cli": "^1.9.0",
```

File: components/PostContent/PostContent.tsx [hunk 1]
```diff
@@ -15,53 +15,54 @@ export default function PostContent( props: {
...
-					const key = `block-${i}`;
+					const defaultProps = { key: `block-${i}` };
 
 					switch ( block.name ) {
 						case 'core/image':
+							const imageProps = {
+								...defaultProps,
+								src: blockProps.src,
+								srcSet: blockProps.srcset,
+								key: blockProps.key,
```
Comment (chriszarate) [inline components/PostContent/PostContent.tsx]: `key` should never be overridden from WordPress data. It's purely a React construct.
```suggestion
```

File: components/PostContent/PostContent.tsx [hunk 2]
```diff
@@ -15,53 +15,54 @@ export default function PostContent( props: {
 			{
 				props.blocks.map( ( block, i ) => {
 					const blockProps = mapAttributesToProps( block.attributes || [] );
-					const key = `block-${i}`;
+					const defaultProps = { key: `block-${i}` };
 
 					switch ( block.name ) {
 						case 'core/image':
+							const imageProps = {
```
Comment (chriszarate) [inline components/PostContent/PostContent.tsx:25]: This isn't really addressing the fact that `srcSet` was removed. All that this is doing is outsmarting the type checker. I'm really not sure how, though. 🤔

File: components/PostContent/PostContent.tsx (hunk 1 above)
Comment (luiztiago) [reply components/PostContent/PostContent.tsx]: @chriszarate Good catch. I miscopied it and just fixed. Sorry

**Review by chriszarate — CHANGES_REQUESTED** [review]
This looks great overall! Two things I'd like to pair on before we merge:

- We need to update our `Image` component to directly utilize the new `deviceSizes` prop. This will be a fun exploration of how WordPress delivers image data and an adventure in responsiveness. :)
- Our platform is not ready to support the image optimization tooling that Next.js now provides, so we need a way to flag that for users.

**Comment by luiztiago** [comment]
@chriszarate - I just updated the next.js version to 12.0.7. So as you can see [here](https://github.com/vercel/next.js/releases/tag/v12.0.5), the upgrade to 12.0.5+ is recommended for all users. A backport of the patch to Next.js 11 is available as 11.1.3.

**Review by chriszarate — APPROVED** [review]
Ok, let's merge this and work on the Image component separately.

**Pull request merged**